src/System.IO.FileSystem.Watcher.Polling/System/IO/PollingFileSystemWatcher.cs

```diff
@@ -131,12 +133,33 @@ internal void UpdateState(string directory, ref FileChangeList changes, ref File
         /// </summary>
         public void Dispose()
         {
-            _timer.Dispose();
+            Dispose(true);
+            GC.SuppressFinalize(this);
         }
 
         public void Dispose(WaitHandle notifyObject)
         {
-            _timer.Dispose(notifyObject);
+            Dispose(true, notifyObject);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing, WaitHandle notifyObject = null)
+        {
+            if (disposed) return;
+
+            if (disposing)
+            {
+                if (notifyObject != null)
+                {
+                    _timer.Dispose(notifyObject);
```

Comment:
Don't do this here. Presumably we always want the timer disposed, even if the derived class doesn't call base.

Reply:
`_timer` is always disposed. It's just switching which `Dispose` method to use. Am I missing something?

Reply:
If a derived class overrides your new virtual Dispose() and never calls .Base() you won't get called. I think it is a good idea to have a virtual Dispose(), we just shouldn't do anything in it in the base class.

Reply:
Uh, so you don't want me to dispose `_timer`?

Reply:
> Uh, so you don't want me to dispose _timer?

No, I don't want you to dispose it in a virtual method. Put the dispose calls back into the non-virtual methods.

Reply:
I was following the pattern here: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#implementing-the-dispose-pattern-for-a-base-class

Reply:
Shouldn't it be up to the developer to properly inherit?

Reply:
> Shouldn't it be up to the developer to properly inherit?

If you can make it more stable, why wouldn't you? Particularly because this one has timed events. Some examples:

1. What if the derived class takes forever and another event fires? Or worse, in rare cases?
2. What if the derived class throws and another event fires because we didn't "unhook"?

Disposals in my experience can go horribly wrong. If we were to expose the timer (e.g. via protected), then I'd think differently in this case. (I'd rather we didn't, btw.)

Reply:
Is this what you are thinking?

```csharp
        public void Dispose()
        {
            _timer.Dispose();
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        public void Dispose(WaitHandle notifyObject)
        {
            _timer.Dispose(notifyObject);
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            disposed = true;
        }
```

Reply:
> Is this what you are thinking?

Closer. If we care about _disposed, we should set it before we call _timer.Dispose() so that we can skip calling _timer.Change(). You might want to consider checking for it in Start() as well- potentially worth throwing ObjectDisposed if that happens.

One possibility is to not add another member and just set _timer to null and check for null where we care.